sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/BeginInvokePairedWithEndInvoke.cs

```diff
@@ -0,0 +1,230 @@
+                    {
+                        var callbackArg = GetCallbackArg(invocation);
+                        if (callbackArg != null &&
+                            (IsNullLiteral(callbackArg) || !IsCallbackMayContainEndInvoke(callbackArg, semantic)) &&
+                            !IsParentMethodContainsEndInvoke(invocation, semantic))
+                        {
+                            var location = ((SyntaxToken)invocation.GetMethodCallIdentifier()).GetLocation();
+                            c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, location));
+                        }
+                    }
+                },
+                SyntaxKind.InvocationExpression);
+        }
+
+        private static bool IsParentMethodContainsEndInvoke(SyntaxNode node, SemanticModel semantic)
+        {
+            var parentContext = node.AncestorsAndSelf()
+                .FirstOrDefault(ancestor => ParentDeclarationKinds.Contains(ancestor.Kind()));
+            return GetEndInvokeList(parentContext, semantic).Count > 0;
+        }
+
+        private ExpressionSyntax GetCallbackArg(InvocationExpressionSyntax invocationExpression)
+        {
+            var callbackArgPos = invocationExpression.ArgumentList.Arguments.Count - 2;
+            var callbackArg = GetArgumentExpressionByNameOrPosition(invocationExpression, "callback", callbackArgPos)
+                ?.RemoveParentheses();
+            return callbackArg;
+        }
+
+        private static bool IsCallbackMayContainEndInvoke(SyntaxNode callbackArg, SemanticModel semantic)
+        {
+            var calledMethod = callbackArg.RemoveParentheses();
+            calledMethod = ReplaceIdentifierByItsInitializer(calledMethod, semantic);
+            calledMethod = ReplaceNewDelegateByMethodDeclaration(calledMethod, semantic);
+            if (calledMethod != null && ParentDeclarationKinds.Contains(calledMethod.Kind()))
+            {
+                return GetEndInvokeList(calledMethod, semantic).Count > 0;
+            }
+
+            return true;
+        }
+
+        private static SyntaxNode ReplaceNewDelegateByMethodDeclaration(SyntaxNode node, SemanticModel semantic)
+        {
+            if (node != null && node.IsKind(SyntaxKind.ObjectCreationExpression))
+            {
+                var arguments = ((ObjectCreationExpressionSyntax)node).ArgumentList.Arguments;
+                if (arguments.Count == 1 &&
+                    semantic.GetSymbolInfo(arguments[0].Expression).Symbol is IMethodSymbol symbol)
+                {
+                    return symbol?.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax();
+                }
+            }
+
+            return node;
+        }
+
+        private static SyntaxNode ReplaceIdentifierByItsInitializer(SyntaxNode node, SemanticModel semantic)
+        {
+            if (node != null && node.IsKind(SyntaxKind.IdentifierName))
```

Comment:
You could move the `IsKind` check in the parent, and make this method just for IdentifierNameSyntax parameter instead of SyntaxNode - like this it would be clear from the signature what it should do